### PR TITLE
[routing-manager] add `Array` & use it for OMR/on-link prefix list

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two related commits:

**[routing-manager] Use `Array` for list of OMR or on-link prefixes**

This commit simplifies the `RoutingManager` class by using `Array`
to store lists of new or advertised OMR/on-link prefixes.

**[common] adding `Array` class**

This commit adds a generic `Array<Type, kMaxSize>` class which
represents a fixed size array of elements of template `Type` which
tracks its current length. Entries can be appended to the array
(`PushBack()`) and removed from the end of array (`PopBack()`). This
simple data model is used in some other modules, so having a generic
version can help simplify the code. This commit also adds a unit test
to verify `Array` methods.

### CODE REVIEW

File: src/core/common/array.hpp [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+     * This method indicates whether or not the array is empty.
+     *
+     * @retval TRUE when array is empty.
+     * @retval FALSE when array is not empty.
+     *
+     */
+    bool IsEmpty(void) const { return (mLength == 0); }
+
+    /**
+     * This method indicates whether or not the array is full
```
Comment (wgtdkp) [inline src/core/common/array.hpp:80]: nit:
```suggestion
     * This method indicates whether or not the array is full.
```

**Review by wgtdkp — APPROVED** [review]
Thanks @abtink ! 👍 
I believe this Class will be useful to many other modules!

File: src/core/common/array.hpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+    /**
+     * This method returns the maximum array size (max number of elements).
+     *
+     * @returns The maximum array size (max number of elements that can be added to the array).
+     */
+    uint16_t GetMaxSize(void) const { return kMaxSize; }
+
+    /**
+     * This method return the current length of array (number of elements).
```
Comment (jwhui) [inline src/core/common/array.hpp:106]:
```suggestion
     * This method returns the current length of array (number of elements).
```

File: src/core/common/array.hpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+     * @retval FALSE when array is not full.
+     *
+     */
+    bool IsFull(void) const { return (mLength == kMaxSize); }
+
+    /**
+     * This method returns the maximum array size (max number of elements).
+     *
+     * @returns The maximum array size (max number of elements that can be added to the array).
+     */
```
Comment (jwhui) [inline src/core/common/array.hpp:102]:
```suggestion
     * @returns The maximum array size (max number of elements that can be added to the array).
     *
     */
```

File: src/core/common/array.hpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+     *
+     * @returns The current array length.
+     *
+     */
+    uint16_t GetLength(void) const { return mLength; }
+
+    /**
+     * This method overloads the `[]` operator to get the element at a given index.
+     *
+     * This method does not perform index bounds checking. Behavior is undefined if the @p aIndex is not valid.
```
Comment (jwhui) [inline src/core/common/array.hpp:116]:
```suggestion
     * This method does not perform index bounds checking. Behavior is undefined if @p aIndex is not valid.
```

File: src/core/common/array.hpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+     *
+     * @returns A reference to the element in array at @p aIndex.
+     *
+     */
+    Type &operator[](uint16_t aIndex) { return mElements[aIndex]; }
+
+    /**
+     * This method overloads the `[]` operator to get the element at a given index.
+     *
+     * This method does not perform index bounds checking. Behavior is undefined if the @p aIndex is not valid.
```
Comment (jwhui) [inline src/core/common/array.hpp:128]:
```suggestion
     * This method does not perform index bounds checking. Behavior is undefined if @p aIndex is not valid.
```

File: tests/unit/test_array.cpp [hunk 6]
```diff
@@ -0,0 +1,255 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
```
Comment (jwhui) [inline tests/unit/test_array.cpp:2]:
```suggestion
 *  Copyright (c) 2021, The OpenThread Authors.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**